Routes/chat

sonamdkindy opened the pull request:
The updates to the frontend were only pushed last night, so I haven't had a chance to look at them yet.

I can't leave a detailed comment about the changes, because I have limited time, but if you run `npm run apidoc` and then `npm start`, you can view details about the routes that I created, the expected success responses and possible error responses.

ernewtoner commented on src/routes/chats.js:
```diff
@@ -1,25 +1,229 @@
+	* @apiHeader {String} token Authorization Bearer Token.
+	*
+	* @apiSuccess {Object[]} - List of chats.
+	* @apiSuccess {String} -.id Chats id.
```
Think this is supposed to be `{Number} id Chat ID.` right?
```suggestion
	* @apiSuccess {Number} id Chats ID.
```

ernewtoner commented on src/routes/chats.js:
```diff
@@ -1,25 +1,229 @@
+	* @apiHeader {String} token Authorization Bearer Token.
+	*
+	* @apiSuccess {Object[]} - List of chats.
+	* @apiSuccess {String} -.id Chats id.
```
Same as above I think.
```suggestion
	* @apiSuccess {Number} id Chats ID.
```

ernewtoner commented on src/controllers/chat.controller.js:
```diff
@@ -0,0 +1,122 @@
+        const message = [
+            req.params.cid,
+            userId,
+            req.body.text
```
I wonder if we want these fields to exactly match up with the Redux state, guess it would make sense to make them mirror each other as much as possible.

sonamdkindy replied:
Manipulating the data so it's in the format expected on the frontend is atypical from what I've heard. Usually the data is retrieved in whatever format from the service on the frontend and the proper transformations are made to the data to suit the frontend/views. If there's anything major that you feel should be changed though, let me know, @mnicole. 

Also, I just corrected some typos/missing info that I made earlier in the apidoc comments - thanks for catching a couple, so I was able to take a quick look again and correct all - I think - of them, @ernewtoner.

sonamdkindy replied:
Also not sure what you mean by these fields specifically? This actually references the db design, which we worked on for the project plan and during the first week of work (week 3), where the message table has columns for chat ID (`chats_id`), user ID (`users_id`) and text (`text`). `created_at` defaults to the present time. Changing the naming conventions at this point would be a bit arduous, but it's possible, if we find that that's seriously necessary.  What would have been better is looking at the db design and using that to format the frontend design...

ernewtoner replied:
Interesting, yeah that makes sense, seems like data is always kind of decoupled like that and doesn't need to match up sequentially in any way.

Ah no worries about the comments--figured you'd catch them anyway once you looked at it again.

ernewtoner replied:
Just saw your 2nd comment (page wasn't refreshed)--was just wondering in general, definitely no need to change the db design. Yeah I'm more thinking about the Redux design, could revisit it in those respects.

sonamdkindy replied:
I thought you might've checked that when you were setting up redux, but I didn't check your PR to verify. It's unfortunate if you find you have to make a lot of modifications. I'm afraid it wouldn't make a lot of sense to format the data in a specific way on the backend contrary to the DB design just for the frontend. It would've made/makes more sense to design the frontend accordingly. That would've been something crucial to look at for the redux design, I'm afraid.

sonamdkindy replied:
If that's something that's a major roadblock, we should discuss/I can help with whatever. This is why communication and looking at all the information available, like the DB design that was already decided upon early on, is super important.

ernewtoner replied:
I did look at the db when doing redux state and it matches it mostly--with maybe the one mistake of using username instead of user id--sorry if my comment was misleading, I was just making a general comment about matching them up cause I was wondering if fields being sequentially matched woulb e a good thing (which we addressed in the above comments), didn't mean to directly comment on or criticize the backend implementation. The Redux state can be modified easily right now, so it's no big deal if we want to make modifications.

sonamdkindy replied:
No worries, I didn't think you were criticizing the backend; I was just responding to your inquiry. It's not just the db design but other aspects of the backend that would have been important to consider. We can discuss more about how to proceed at the meeting today, so we can avoid this sort of lapse in communication and design in the future.

sonamdkindy commented:
I think that's all of the typos now fixed. JFC, that was annoying. 

I did my best with the routes given our DB design and timing constraints with the frontend, my schedule, etc. Plz let me know if you see anything else, or if there's anything you think we should change on the server-side.

sonamdkindy commented on src/controllers/chat.controller.js:
```diff
@@ -0,0 +1,124 @@
+        return error.constraint === 'users_chats_pkey' ? 
+            Promise.reject('User already member of chat') :
+            Promise.reject(error);
+    }
```
This is just to be extra careful that we only specify the user was already added to the chat/is already a member of the chat if the primary key constraint was violated. Otherwise, whatever error due to incorrect values provided in the prepared statement will be thrown instead.

sonamdkindy commented:
FYI, the dev't database is now seeded as part of running `npm start`, i.e. whenever you run `npm start`, it'll run `npm prestart` automatically and that will set up the database schema (wipe data previously added) and seed the database with test data. I can pretty easily make adjustments to the test data created, so let me know if any particular changes would be helpful, @mnicole and @ernewtoner.

sonamdkindy commented:
We discussed these changes in our meeting earlier this afternoon, but just to recap:
- make login and signup routes case-insensitive for the email provided (I also updated the test to verify this was working as expected)
- `GET /user` by email return id so we can make add user request for the chat (I also added tests)
- `GET /chat` should be a list of chats that includes an array of users including their id and display_name (This took some finagling to do with some finesse; I didn't want to have to iterate across the chat IDs and across each one across every user ID, which is what a relatively simple query would require - I ended up using a *sick* PostgreSQL method called `json_agg` to aggregate all of the users into an array of objects (and `json_build_object` to create the appropriate property names for the aggregate) and grouped them by the chat id + name
- I also updated all of the comments for apidoc where applicable so that auto-generated documentation should reflect everything that is new/has been tweaked

sonamdkindy commented:
Updated tests with more specificity and in turn noticed I had forgotten to add the user ID to the user obj returned from `GET /chat/:cid`. All good now. Merging!
